### PR TITLE
Change Window Theme with Pulsar Theme

### LINKED ISSUE #422: Inconsistent theme on the top bar
### Thanks in advance for your bug report!

- [X] Have you reproduced issue in [safe mode](https://pulsar-edit.dev/docs/atom-archive/faq/sections/what-does-safe-mode-do.html)?
- [X] Have you used the [debugging guide](https://pulsar-edit.dev/docs/atom-archive/hacking-atom/#debugging) to try to resolve the issue?
- [X] Have you checked our [FAQs](https://pulsar-edit.dev/docs/launch-manual/sections/faq/) to make sure your question isn't answered there?
- [X] Have you checked to make sure your issue does not already [exist](https://github.com/issues?q=is%3Aissue+user%3Apulsar-edit+is%3Aopen)?
- [X] Have you checked you are on the latest release of Pulsar?

### What happened?

On GNU Linux (KDE and Gnome) the top bar 
image
does not change making it inconsistent.

### Pulsar version

102.0

### Which OS does this happen on?

🐧 Red Hat based (Fedora, Alma, RockyLinux, CentOS Stream, etc.)

### OS details

Fedora 37

### Which CPU architecture are you running this on?

64-bit(x86_64)

### What steps are needed to reproduce this?

1. Change the system theme to dark mode
2. Change the Pulsar theme to any, can more easily be seen on dark mode.
3. Notice that the top bar doesn't change

### Additional Information:

image
image

### PR DESCRIPTION
**Pull request by confused-Techie**
This PR adds a feature that's likely long overdue.

When a user changes the theme within Pulsar, the brightness of the background color will be computed, and if light or dark, will change the Pulsar Window theme to respect that. During initial boot the theme will still follow the system theme as it does, but otherwise once themes are loaded it will then be changed if needed.

## How is this done

When the `theme-manager` enables a new theme, it will use the `color.js` class to get the RGB values of the `HTML` element's `backgroundColor`, with that it then uses the formula defined by [`ITU-R BT.709`](https://www.itu.int/dms_pubrec/itu-r/rec/bt/R-REC-BT.709-6-201506-I!!PDF-E.pdf) to determine if that colour is brighter or darker. 

Depending on that answer it will then call `application-delegate.js` (which is now provided to the Theme Manager) `.setWindowTheme()` which uses the IPC Bus to trigger the [`nativeTheme` module](https://www.electronjs.org/docs/latest/tutorial/dark-mode) from Electron to change the theme.

## Issues 

According to [this](https://github.com/electron/electron/issues/23479) as far as I can tell this module does not respect the theme for Windows on the title bar, but does so for the menu bar. So this may only half way fix the issue in Windows until we could either bump Electron versions, or someone that understands that discussion better can implement it. But I'm hoping this does work properly for users on other platforms.

## Examples 

Again remember, that an Electron issue prevents this working on the Title bar in Windows, which is where I took these pictures. So remembering that here's the results I was able to acheive.

### When set to a light theme:

image

### When set to a dark theme: 
image

---

Resolves #422

### CODE REVIEW

File: src/theme-manager.js [hunk 1]
```diff
@@ -369,6 +372,20 @@ On linux there are currently problems with watch sizes. See
...
   }
 
+  refreshWindowTheme() {
+    let bgColor = Color.parse(getComputedStyle(document.documentElement).backgroundColor);
+
+    let luminosity = 0.2126 * bgColor.red + 0.7152 * bgColor.green + 0.0722 * bgColor.blue;
+    // ^^ Luminosity per ITU-R BT.709
+    if (luminosity < 40) {
+      // Considered Dark
+      this.applicationDelegate.setWindowTheme({ setTheme: "dark" });
```
Comment (savetheclocktower) [inline src/theme-manager.js:382]: Any reason this isn't something like

```js
this.applicationDelegate.setWindowTheme('dark');
```

instead?

Comment (confused-Techie) [reply src/theme-manager.js:382]: None at all really. That object structure comes when in my first draft I used `windowTheme()` and was wanting to see if I could do `setTheme: "", getTheme: true`

But after getting everything working, additional APIs weren't needed, so I scrapped them and never changed the object to something simpler. So I'll make sure to do so

Comment (confused-Techie) [reply src/theme-manager.js]: Finally took this suggestion to heart. Thanks again for it

**Comment by confused-Techie** [comment]
Made the requested changes on this PR.

While it doesn't resolve all the issues related, do we think we should go ahead and close this one? Or still worth it to implement?

**Comment by Spiker985** [comment]
Provided this PR still works in newer versions, and doesn't cause problems - we should probably implement this, no? It doesn't fix _everything_ with the weird theming, but it does fix _some_, right?

**Comment by savetheclocktower** [comment]
It _might_ be worth putting behind a config setting, I suppose? I can see how it makes sense to infer a light-mode or dark-mode decision based on the UI theme the user has chosen… but I can also see how some people might just want all their window title bars to match each other.

**Comment by Spiker985** [comment]
That's an easy agreement from me - especially since it would technically be a user facing change, and something that would be a divergence from their expected norm

**Comment by confused-Techie** [comment]
Thanks for taking another look at this @Spiker985 & @savetheclocktower I'll go ahead and put the behavior behind a config file and get this updated as soon as I can

**Comment by confused-Techie** [comment]
Alright, @Spiker985 & @savetheclocktower I've gone ahead and made sure this one is configurable. If anyone would like to review it, it'd be awesome to get this one merged

**Review by DeeDeeG — APPROVED** [review]
Seems legit. 👍 

Maybe someday we can add something to auto-select a light or dark theme based on the OS theme. (Perhaps with a drop-down for system dark and another dropdown for system light, to designate the specific theme when OS theme is light vs when OS theme is dark.)

**Comment by confused-Techie** [comment]
Thanks for the review @DeeDeeG!
I'll go ahead and merge this one, and you're sources are basically correct for where I got my information.

As for syncing this setting with the OS theme, that should be the default behavior once we bump Electron versions (iirc) so hopefully that's a non-issue in the future.

**Pull request merged**